File: Bikesharing/src/simulator/parameters.py
class Solver_Parameters:
    """
    Encapsulates selected model parameters for optimization and solution approaches.

    Attributes
    ----------
    epoch_length : int
        Epoch length for re-optimization.
    nb_threads : int
        Number of threads for parallel optimization.
    main_algorithm : MainAlgorithm
        The algorithm used for solving the master problem.
    nb_column : int
        Number of columns to be added in column generation.
    is_truncated : bool
        If a truncated labeling strategy is used.
    max_label : int
        Maximum number of labels to keep while using truncated labeling.
    is_drop_pick_possible : bool
        If picking up after dropping is allowed as an acceleration strategy.
    labeling_strategy : LabelingStrategy
        Labeling strategy used.
    nb_pick : int
        Number of pickups allowed.
    sort_path : SortPaths
        Sorting mode of paths.
    mip_gap : float
        MIP gap used as a stopping criterion.
    """

    def __init__(self, epoch_length, nb_threads, main_algorithm, nb_column, is_truncated,
                 max_label, is_drop_pick_possible, labeling_strategy, nb_pick, sort_path, mip_gap):
        self.epoch_length = epoch_length
        self.nb_threads = nb_threads
        self.main_algorithm = main_algorithm
        self.nb_column = nb_column
        self.is_truncated = is_truncated
        self.max_label = max_label
        self.is_drop_pick_possible = is_drop_pick_possible
        self.labeling_strategy = labeling_strategy
        self.nb_pick = nb_pick
        self.sort_path = sort_path
        self.mip_gap = mip_gap

    def __str__(self):
        """
        Generates a string representation of the Parameters object, detailing all relevant attributes dynamically.
        """
        class_string = f"{self.__class__.__name__}: {{"
        for attribute, value in self.__dict__.items():
            attr_name = attribute.lstrip('_')
            class_string += f"{attr_name}: {value}, "
        class_string = class_string.rstrip(', ') + "}"
        return class_string

class Parameters:
    """
    Encapsulates simulation parameters for a transportation network.

    Attributes
    ----------
    simulation_start : int
        The start time of the simulation in seconds since the beginning of the day.
    num_vehicles : int
        The number of vehicles available in the simulation.
    num_stations : int
        The number of stations in the simulation.
    time_per_bike : float
        The average time (in seconds) it takes to handle each bike.
    time_to_park : int
        The time (in seconds) it takes to park a vehicle at a station.
    """

    def __init__(self, simulation_start, num_vehicles, num_stations, time_per_bike, time_to_park):
        self.simulation_start = simulation_start
        self.num_vehicles = num_vehicles
        self.num_stations = num_stations
        self.time_per_bike = time_per_bike
        self.time_to_park = time_to_park

    def __str__(self):
        """
        Generates a string representation of the SimulationParameters object.
        """
        return (f"SimulationParameters(simulation_start={self.simulation_start}, "
                f"num_vehicles={self.num_vehicles}, num_stations={self.num_stations}, "
                f"time_per_bike={self.time_per_bike}, time_to_park={self.time_to_park})")

File: Bike_Sharing/src/data/Parameters.cpp
//
// Created by Elahe Amiri on 2022-10-13.
//

#include "Parameters.h"

//************************************************************************
//                     FUNCTIONS FOR PARAMETERS CLASS
//************************************************************************

// Constructor and Destructor
Parameters::Parameters(int epochLength, int nbThreads, MainAlgorithm mainAlgorithm, bool oneIter, bool greedyReOptimize,
                       int nbColumn, bool isTruncated, int maxLabel, bool isDropPickPossible, int nbStop,
                       SortPaths sortPath, float MIPGap) :
        epochLength_(epochLength), nbThreads_(nbThreads), mainAlgorithm_(mainAlgorithm),
        oneIter_(oneIter), greedyReOptimize_(greedyReOptimize),
        nbColumn_(nbColumn), isTruncated_(isTruncated), MaxLabel_(maxLabel), isDropPickPossible_(isDropPickPossible),
        nbStop_(nbStop), sortPath_(sortPath),MIPGap_(MIPGap) {}

Parameters::~Parameters() = default;

// Display function
std::string Parameters::toString() const {
    std::stringstream repStr;
    repStr << "# MODEL PARAMETERS" << std::endl;
    repStr << "#" << std::endl;
    int setwLength = 30;
    repStr << std::left << std::fixed << std::setprecision(1) << std::boolalpha;
    repStr << std::setw(setwLength) << "# epoch Length " << " = " << epochLength_ << " (s)" << std::endl;
    repStr << std::setw(setwLength) << "# number of threads " << " = " << nbThreads_ << std::endl;
    repStr << std::setw(setwLength) << "# main algorithm " << " = " << mainAlgorithmName[mainAlgorithm_] << std::endl;
    repStr << std::setw(setwLength) << "# One iter per epoch " << " = " << oneIter_ << std::endl;
    repStr << std::setw(setwLength) << "# Is Greedy Re-Optimized " << " = " << greedyReOptimize_ << std::endl;
    repStr << std::setw(setwLength) << "# Column added to MP " << " = " << nbColumn_ << std::endl;
    repStr << std::endl;

    repStr << "# LABEL SETTING STRATEGIES" << std::endl;
    repStr << "#" << std::endl;
    repStr << std::setw(setwLength) << "# Use Truncated Labeling " << " = " << isTruncated_ << std::endl;
    repStr << std::setw(setwLength) << "# MaxLabel in Truncating " << " = " << MaxLabel_ << std::endl;
    repStr << std::setw(setwLength) << "# Is Pickup allowed after Drop " << " = " << isDropPickPossible_ << std::endl;
    repStr << std::setw(setwLength) << "# number of pickups allowed " << " = " << nbStop_ << std::endl;
    repStr << std::setw(setwLength) << "# Sorting mode of paths " << " = " << SortPathsName[sortPath_] << std::endl;
    repStr << std::endl;

    repStr << "# CPLEX PARAMETERS" << std::endl;
    repStr << "#" << std::endl;
    repStr << std::left << std::fixed << std::setprecision(0);
    repStr << std::setw(setwLength) << "# MIP Gap " << " = " << MIPGap_ << std::endl;

    return repStr.str();

}

std::string Parameters::toStr() const {
    std::stringstream repStr;
    repStr << epochLength_ << ",";
    repStr << nbThreads_ << ",";
    repStr << mainAlgorithmName[mainAlgorithm_] << ",";
    repStr << boolToString(oneIter_) << ",";
    repStr << boolToString(greedyReOptimize_) << ",";
    repStr << nbColumn_ << ",";
    repStr << boolToString(isTruncated_) << ",";
    repStr << MaxLabel_ << ",";
    repStr << boolToString(isDropPickPossible_) << ",";
    repStr << nbStop_ << ",";
    repStr << SortPathsName[sortPath_] << ",";
    repStr << MIPGap_ << "\n";
    return repStr.str();
}


//-----------------------------------------------------------------------------
//  Solver Option Struct
//-----------------------------------------------------------------------------

solverOption::solverOption(bool isTruncated, int maxLabel, bool isDominanceReleased, int nbStop, SortPaths pathSort,
                           bool isDropPickPossible) :
        isTruncated_(isTruncated), nbStop_(nbStop), MaxLabel_(maxLabel), pathSort_(pathSort),
        isDropPickPossible_(isDropPickPossible){}

solverOption::~solverOption() = default;


solverOption::solverOption(PParameters &MainParams) {
    isTruncated_ = MainParams->isTruncated_;
    MaxLabel_ = MainParams->MaxLabel_;
    isDropPickPossible_ = MainParams->isDropPickPossible_;
    nbStop_ = MainParams->nbStop_;
    pathSort_ = MainParams->sortPath_;
}


std::string boolToString(bool value) {
    return value ? "True" : "False";
}
